gnovm/pkg/gnolang/type_check.go

```diff
@@ -215,6 +215,38 @@ func assertAssignableTo(xt, dt Type, autoNative bool) {
 	}
 }
 
+// checkValConstType checks the type and values are valid for a ValueDecl of const.
+func checkValConstType(d *ValueDecl) {
+	if d.Type != nil {
+		switch d.Type.(type) {
+		case *BasicLitExpr, *NameExpr:
+			// Valid constant type expression, in case of NameExpr should evaluate if underlying type is a basic type
+		default:
+			panic("const type should be a basic type")
+		}
+	}
+	for _, vx := range d.Values {
+		checkValConstValue(vx)
+	}
+}
+
+// checkValConstValue validate an expr is valid as a value for a const
+func checkValConstValue(expr Expr) {
+	switch x := expr.(type) {
+	case *BasicLitExpr, *NameExpr:
+		// Valid constant type expression, in case of NameExpr should evaluate if underlying type is a basic type
+	case *BinaryExpr:
+		checkValConstValue(x.Left)
+		checkValConstValue(x.Right)
+	case *UnaryExpr:
+		checkValConstValue(x.X)
+	case *CallExpr:
+		checkValConstValue(x.Func)
```

Comment:
see this case:
```go
package main

type f func()

func foo() {
	println("foo")
}

const ff = foo

func main() {
	println("ok")
}
```

it should be excluded, right?

Reply:
Thanks for comment, In fact, I'm having a bit of trouble figuring out how to check that a name expr is a const or a cast of a basic type, for example `const x = uint(8)`, do you have any tips to give me on this? In the meantime, I'll keep working on the code and keep looking, thanks.

Reply:
It seems there may be a misunderstanding about the use of nameExpr as the right-hand side (RHS) of a const declaration. For example:

go
Copy code
const a = 1
// 'a' qualifies as a ConstExpr
const b = a
In this case, using a as the RHS in the declaration of b is perfectly valid since a is evaluated to be a  constant expression.

Regarding the expression uint(8), this is a CallExpr with the function type(TypeType) of uint8. I'm not certain if this directly addresses your question. 

However, I recommend we ensure a shared understanding of the related concepts by reviewing the section on constants in the Go language specification: [Go Constants](https://go.dev/ref/spec#Constants).


Tips, you can try to conduct `checkValConstValue` in `trans_leave *ValueDecl` to see if it works.